<a href="https://colab.research.google.com/github/MaggieWelch/MLProject/blob/main/MLProject.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Exploring data

In [53]:
import pandas as pd

df = pd.read_csv('Spotify Most Streamed Songs.csv')
print(df.head())
df.shape

                            track_name    artist(s)_name  artist_count  \
0  Seven (feat. Latto) (Explicit Ver.)  Latto, Jung Kook             2   
1                                 LALA       Myke Towers             1   
2                              vampire    Olivia Rodrigo             1   
3                         Cruel Summer      Taylor Swift             1   
4                       WHERE SHE GOES         Bad Bunny             1   

   released_year  released_month  released_day  in_spotify_playlists  \
0           2023               7            14                   553   
1           2023               3            23                  1474   
2           2023               6            30                  1397   
3           2019               8            23                  7858   
4           2023               5            18                  3133   

   in_spotify_charts    streams  in_apple_playlists  ...  key   mode  \
0                147  141381703                  4

(953, 25)

In [54]:
print(df.columns.tolist())

['track_name', 'artist(s)_name', 'artist_count', 'released_year', 'released_month', 'released_day', 'in_spotify_playlists', 'in_spotify_charts', 'streams', 'in_apple_playlists', 'in_apple_charts', 'in_deezer_playlists', 'in_deezer_charts', 'in_shazam_charts', 'bpm', 'key', 'mode', 'danceability_%', 'valence_%', 'energy_%', 'acousticness_%', 'instrumentalness_%', 'liveness_%', 'speechiness_%', 'cover_url']


Dropping variables/columns that are not relevent to our analysis

In [55]:
columns_to_drop = ['artist_count', 'released_month', 'released_day', 'in_spotify_playlists', 'in_spotify_charts', 'in_apple_playlists', 'in_apple_charts', 'in_deezer_playlists', 'in_deezer_charts', 'in_shazam_charts', 'cover_url', ]
df_cleaned = df.drop(columns=columns_to_drop)
df_cleaned.head()

track_name    artist(s)_name  released_year  \
0  Seven (feat. Latto) (Explicit Ver.)  Latto, Jung Kook           2023   
1                                 LALA       Myke Towers           2023   
2                              vampire    Olivia Rodrigo           2023   
3                         Cruel Summer      Taylor Swift           2019   
4                       WHERE SHE GOES         Bad Bunny           2023   

     streams  bpm key   mode  danceability_%  valence_%  energy_%  \
0  141381703  125   B  Major              80         89        83   
1  133716286   92  C#  Major              71         61        74   
2  140003974  138   F  Major              51         32        53   
3  800840817  170   A  Major              55         58        72   
4  303236322  144   A  Minor              65         23        80   

   acousticness_%  instrumentalness_%  liveness_%  speechiness_%  
0              31                   0           8              4  
1               7                   0          10              4  
2              17                   0          31              6  
3              11                   0          11             15  
4              14                  63          11              6

Converting the streams variable values into numerical values, and then attempting to organize them numerically

In [56]:
df['streams'] = pd.to_numeric(df['streams'], errors='coerce')
df.sort_values(by='streams')

track_name  \
123                          Que Vuelvas   
393                    Jhoome Jo Pathaan   
144                                QUEMA   
142          Gol Bolinha, Gol Quadrado 2   
68                             Overdrive   
..                                   ...   
620                         Dance Monkey   
86                     Someone You Loved   
179                         Shape of You   
55                       Blinding Lights   
574  Love Grows (Where My Rosemary Goes)   

                                        artist(s)_name  artist_count  \
123                         Carin Leon, Grupo Frontera             2   
393  Arijit Singh, Vishal Dadlani, Sukriti Kakar, V...             6   
144                       Sog, Ryan Castro, Peso Pluma             3   
142                                Mc Pedrinho, DJ 900             2   
68                                         Post Malone             1   
..                                                 ...           ...   
620                                        Tones and I             1   
86                                       Lewis Capaldi             1   
179                                         Ed Sheeran             1   
55                                          The Weeknd             1   
574                                  Edison Lighthouse             1   

     released_year  released_month  released_day  in_spotify_playlists  \
123           2022              12             9                   763   
393           2022              12            22                   138   
144           2023               7            13                   437   
142           2023               6             1                   293   
68            2023               7            14                   410   
..             ...             ...           ...                   ...   
620           2019               5            10                 24529   
86            2018              11             8                 17836   
179           2017               1             6                 32181   
55            2019              11            29                 43899   
574           1970               1             1                  2877   

     in_spotify_charts       streams  in_apple_playlists  ...  key   mode  \
123                 26  2.762000e+03                  21  ...   A#  Major   
393                  4  1.365184e+06                  13  ...    G  Major   
144                 31  1.159939e+07                  17  ...  NaN  Major   
142                  8  1.195664e+07                   5  ...    B  Minor   
68                  36  1.478042e+07                  36  ...   C#  Major   
..                 ...           ...                 ...  ...  ...    ...   
620                  0  2.864792e+09                 533  ...   F#  Minor   
86                  53  2.887242e+09                 440  ...   C#  Major   
179                 10  3.562544e+09                  33  ...   C#  Minor   
55                  69  3.703895e+09                 672  ...   C#  Major   
574                  0           NaN                  16  ...    A  Major   

     danceability_% valence_%  energy_% acousticness_% instrumentalness_%  \
123              49        78        64             19                  0   
393              82        62        74             10                  0   
144              79        92        89              5                  0   
142              93        68        65             42                  0   
68               56        48        73              0                  0   
..              ...       ...       ...            ...                ...   
620              82        54        59             69                  0   
86               50        45        41             75                  0   
179              83        93        65             58                  0   
55               50        38        80             

Creating bins

In [57]:
bins = [80000000, 100000000, 120000000, 140000000, df['streams'].max()]

labels = ['Low', 'Medium', 'High', 'Very High']

# Create a new column 'streams_category' based on the bins
df['streams_category'] = pd.cut(df['streams'], bins=bins, labels=labels, include_lowest=True)

print(df[['streams', 'streams_category']].head())



         streams streams_category
933  253650850.0        Very High
934  223064273.0        Very High
935  170413877.0        Very High
936  218320587.0        Very High
937  141720999.0        Very High
938  187339835.0        Very High
939  198365537.0        Very High
940  186104310.0        Very High
941  177503916.0        Very High
942   71095708.0              NaN
943  112436403.0           Medium
944  203436468.0        Very High
945   93367537.0              Low
946  106249219.0           Medium
947  181382590.0        Very High
948   91473363.0              Low
949  121871870.0             High
950   73513683.0              NaN
951  133895612.0             High
952   96007391.0              Low


In [41]:
df['streams_category'].isna().sum()

113

Our initial code that organized streams into bins still has 113 nan values, and so we need to adjust the organization of the bins so that it includes all values. This line of code was generated from chatGBT, and allows us to look at the values that were omitted from the bins.

In [44]:
nan_streams_values = df.loc[df['streams_category'].isna(), 'streams']
print(nan_streams_values)

7      58149378.0
11     58255150.0
17     30546883.0
21     52135248.0
25     78300654.0
          ...    
835    79095270.0
913    47093942.0
916    53987404.0
942    71095708.0
950    73513683.0
Name: streams, Length: 113, dtype: float64
